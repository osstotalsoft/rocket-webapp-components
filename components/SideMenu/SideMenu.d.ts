export interface Props {
  /**
    * The content of the menu
    */
  content?: React.ReactNode
  /**
    * The icon to be displayed on the button
    */
  icon?: Object
  /**
    * Override or extend the styles applied to the side-menu
    */
  customSideMenuClass?: string
  /**
    * Override or extend the styles applied to the button
    */
  customButtonClass?: string
  /**
    * Override or extend the styles applied to the icon
    */
  customIconClass?: string
  /**
    * Override or extend the styles applied to the dropdown
    */
  customContainerClass?: string
}

export type SideMenuProps = Props

export default function SideMenu(props: SideMenuProps): JSX.Element