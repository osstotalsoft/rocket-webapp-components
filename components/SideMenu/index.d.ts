export { default } from './SideMenu'
export * from './SideMenu'