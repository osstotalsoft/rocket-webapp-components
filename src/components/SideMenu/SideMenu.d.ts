export interface Props {
  /**
    * The content of the menu
    */
  content?: React.ReactNode
  /**
    * The icon to be displayed on the button
    */
  icon?: Object
  /**
    * Override or extend the styles applied to the side-menu
    */
  customSideMenuClass?: string
  /**
    * Override or extend the styles applied to the button
    */
  customButtonClass?: string
  /**
    * Override or extend the styles applied to the icon
    */
  customIconClass?: string
  /**
    * Override or extend the styles applied to the dropdown
    */
  customContainerClass?: string
  /**
   * Demos:
   * 
   * - https://bit.dev/totalsoft_oss/react-mui/side-menu
   * 
   *  Take a look over the code for further information: https://github.com/osstotalsoft/rocket-webapp-components/tree/master/components/SideMenu
   */
}

export type SideMenuProps = Props

export default function SideMenu(props: SideMenuProps): JSX.Element