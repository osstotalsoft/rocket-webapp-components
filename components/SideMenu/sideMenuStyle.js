import styles from '../../assets/jss/styles'
import {theme as defaultTheme} from '../../themes/defaultTheme'

const sideMenuStyle = theme => {
  const { whiteColor, grayColor } = styles(theme)

  return {
    sideMenu: {
      position: 'fixed',
      top: '180px',
      right: 0,
      width: '44px',
      height: '38px',
      background: 'rgba(0, 0, 0, 0.3)',
      zIndex: 1031,
      borderRadius: '8px 0 0 8px',
      textAlign: 'center'
    },
    moreButton: {
      display: 'flex',
      justifyContent: 'center',
      height: '100%',
      '&:hover': {
        outline: 'none',
        backgroundColor: theme?.palette?.sideMenu?.hoverBgColor || defaultTheme?.palette?.sideMenu?.hoverBgColor,
        boxShadow: 'none',
        cursor: 'pointer'
      }
    },
    sideMenuIcon: {
      color: `${whiteColor}`,
      height: '100%'
    },
    menuContainer: {
      position: 'absolute',
      top: '8px',
      right: '60px',
      zIndex: 1000,
      display: 'none',
      float: 'left',
      minWidth: '160px',
      minHeight: '40px',
      backgroundColor: `${whiteColor}`,
      opacity: 0,
      border: '1px solid #ccc',
      borderRadius: '4px',
      WebkitBoxShadow: '0 6px 12px rgba(0, 0, 0, 0.175)',
      WebkitTransform: 'translateY(-15%)',
      transformOrigin: '0 0',
      margin: '5px',
      padding: '10px',
      '&:before, &:after': {
        content: '""',
        display: 'inline-block',
        position: 'absolute',
        top: '8%'
      },
      '&:before': {
        borderBottom: '16px solid transparent',
        borderLeft: `16px solid ${grayColor}`,
        borderTop: '16px solid transparent',
        right: '-17px'
      },
      '&:after': {
        borderBottom: '16px solid transparent',
        borderLeft: `16px solid ${whiteColor}`,
        borderTop: '16px solid transparent',
        right: '-16px'
      }
    },
    menuShow: {
      opacity: 1,
      WebkitTransform: 'translateY(-13%)',
      transformOrigin: '0 0',
      display: 'block'
    }
  }
}

export default sideMenuStyle
