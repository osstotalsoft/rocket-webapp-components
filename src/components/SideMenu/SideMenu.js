import React, { useCallback, useEffect, useRef, useState } from 'react'
import cx from 'classnames'
import PropTypes from 'prop-types'
import MenuOpenIcon from '@material-ui/icons/MenuOpen'
import { makeStyles } from '@material-ui/core'
import sideMenuStyle from './sideMenuStyle'

const useStyles = makeStyles(sideMenuStyle)

const SideMenu = ({ icon: Icon, content, customSideMenuClass, customButtonClass, customIconClass, customContainerClass }) => {
  const classes = useStyles()

  const sideMenuClass =
    classes.sideMenu +
    ' ' +
    cx({
      [customSideMenuClass]: customSideMenuClass
    })

  const buttonClass =
    classes.moreButton +
    ' ' +
    cx({
      [customButtonClass]: customButtonClass
    })

  const iconClass =
    classes.sideMenuIcon +
    ' ' +
    cx({
      [customIconClass]: customIconClass
    })

  const menuContainerClass =
    classes.menuContainer +
    ' ' +
    cx({
      [customContainerClass]: customContainerClass
    })

  const [show, setShow] = useState(false)
  const sideMenuRef = useRef()

  const handleClickOutside = useCallback(e => {
    if (sideMenuRef.current && !sideMenuRef.current.contains(e.target)) {
      setShow(false)
    }
  }, [])

  const handleButtonClick = useCallback(() => {
    setShow(current => !current)
  }, [])

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, []) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <div ref={sideMenuRef} className={sideMenuClass}>
      <div onClick={handleButtonClick} className={buttonClass}>
        <Icon fontSize='small' className={iconClass} />
      </div>
      <div className={show ? `${menuContainerClass} ${classes.menuShow}` : menuContainerClass}>{content}</div>
    </div>
  )
}

SideMenu.propTypes = {
  /**
   * The content of the menu
   */
  content: PropTypes.node,
  /**
   * The icon to be displayed on the button
   */
  icon: PropTypes.object,
  /**
   * Override or extend the styles applied to the side-menu
   */
  customSideMenuClass: PropTypes.string,
  /**
   * Override or extend the styles applied to the button
   */
  customButtonClass: PropTypes.string,
  /**
   * Override or extend the styles applied to the icon
   */
  customIconClass: PropTypes.string,
  /**
   * Override or extend the styles applied to the dropdown
   */
  customContainerClass: PropTypes.string
}

SideMenu.defaultProps = {
  icon: MenuOpenIcon
}

export default SideMenu
