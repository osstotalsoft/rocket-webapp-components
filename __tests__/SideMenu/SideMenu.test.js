import React from "react";
import renderer from "react-test-renderer";
import SideMenu from "../../components/SideMenu/SideMenu";
import MenuOpenIcon from '@material-ui/icons/MenuOpen';

describe("SideMenu", () => {
  it("Snapshot", () => {
    const component = renderer.create(
      <SideMenu
        icon={MenuOpenIcon}
        content={"Content of the side-menu"}
      />
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
