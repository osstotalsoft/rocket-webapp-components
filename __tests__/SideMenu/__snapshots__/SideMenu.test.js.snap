// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu Snapshot 1`] = `
<div
  className="makeStyles-sideMenu-1 "
>
  <div
    className="makeStyles-moreButton-2 "
    onClick={[Function]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-sideMenuIcon-3  MuiSvgIcon-fontSizeSmall"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 18h13v-2H3v2zm0-5h10v-2H3v2zm0-7v2h13V6H3zm18 9.59L17.42 12 21 8.41 19.59 7l-5 5 5 5L21 15.59z"
      />
    </svg>
  </div>
  <div
    className="makeStyles-menuContainer-4 "
  >
    Content of the side-menu
  </div>
</div>
`;
